cdap-ui/app/cdap/components/PluginJSONCreator/Create/index.tsx

```diff
@@ -45,158 +40,222 @@ const styles = (): StyleRules => {
+    [configurationGroups, groupToInfo]
+  );
+
+  const widgetContextValue = React.useMemo(
+    () => ({
       groupToWidgets,
+      setGroupToWidgets,
       widgetInfo,
+      setWidgetInfo,
       widgetToAttributes,
+      setWidgetToAttributes,
+    }),
+    [groupToWidgets, widgetInfo, widgetToAttributes]
+  );
+
+  const outputContextValue = React.useMemo(
+    () => ({
       outputName,
+      setOutputName,
+    }),
+    [outputName]
+  );
+
+  const filterContextValue = React.useMemo(
+    () => ({
       filters,
+      setFilters,
       filterToName,
+      setFilterToName,
       filterToCondition,
+      setFilterToCondition,
       filterToShowList,
+      setFilterToShowList,
       showToInfo,
-    });
-  };
+      setShowToInfo,
+    }),
+    [filters, filterToName, filterToCondition, filterToShowList, showToInfo]
+  );
+  const appInternalContextValue = React.useMemo(
+    () => ({
+      activeStep,
+      setActiveStep,
+      liveView,
+      setLiveView,
+      JSONStatus,
+      setJSONStatus,
+    }),
+    [activeStep, setActiveStep, liveView, setLiveView, JSONStatus, setJSONStatus]
+  );
 
-  public state = {
-    activeStep: 0,
-    pluginName: '',
-    pluginType: '',
-    displayName: '',
-    emitAlerts: false,
-    emitErrors: false,
-    configurationGroups: List<string>(),
-    groupToInfo: Map<string, Map<string, string>>(),
-    groupToWidgets: Map<string, List<string>>(),
-    widgetInfo: Map<string, Map<string, string>>(),
-    widgetToAttributes: Map<string, Map<string, any>>(),
-    liveView: true,
-    outputName: '',
-    JSONStatus: JSONStatusMessage.Normal,
-    filters: List(),
-    filterToName: Map({}),
-    filterToCondition: Map<string, Map<string, string>>(),
-    filterToShowList: Map<string, List<string>>(),
-    showToInfo: Map<string, Map<string, string>>(),
-
-    setActiveStep: this.setActiveStep,
-    setBasicPluginInfo: this.setBasicPluginInfo,
-    setConfigurationGroups: this.setConfigurationGroups,
-    setGroupToInfo: this.setGroupToInfo,
-    setGroupToWidgets: this.setGroupToWidgets,
-    setWidgetInfo: this.setWidgetInfo,
-    setWidgetToAttributes: this.setWidgetToAttributes,
-    setLiveView: this.setLiveView,
-    setOutputName: this.setOutputName,
-    setPluginState: this.setPluginState,
-    setJSONStatus: this.setJSONStatus,
-    setFilters: this.setFilters,
-    setFilterToName: this.setFilterToName,
-    setFilterToCondition: this.setFilterToCondition,
-    setFilterToShowList: this.setFilterToShowList,
-    setShowToInfo: this.setShowToInfo,
-  };
+  return (
+    // TODO: Separate providers
+    <PluginInfoContext.Provider value={pluginInfoContextValue}>
+      <ConfigurationGroupContext.Provider value={configuratioGroupContextValue}>
+        <WidgetContext.Provider value={widgetContextValue}>
+          <OutputContext.Provider value={outputContextValue}>
+            <FilterContext.Provider value={filterContextValue}>
```

Comment:
can this nested structure be flatten out? for instance the OutputContext does not really relate to the ConfigurationGroupContext, right?

Reply:
Resolved